pkg/volume/storageos/storageos_util.go

```diff
@@ -0,0 +1,433 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storageos
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"strings"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/volume"
+
+	"github.com/golang/glog"
+	storageosapi "github.com/storageos/go-api"
+	storageostypes "github.com/storageos/go-api/types"
+)
+
+const (
+	losetupPath   = "losetup"
+	apiSecretName = "storageos-api"
+
+	modeBlock deviceType = iota
+	modeFile
+	modeUnsupported
+
+	ErrDeviceNotFound     = "device not found"
+	ErrDeviceNotSupported = "device not supported"
+	ErrNotAvailable       = "not available"
+)
+
+type deviceType int
+
+// storageosVolume describes a provisioned volume
+type storageosVolume struct {
+	ID          string
+	Name        string
+	Namespace   string
+	Description string
+	Pool        string
+	Size        int
+	Labels      map[string]string
+	FSType      string
+}
+
+type storageosAPIConfig struct {
+	apiAddr    string
+	apiUser    string
+	apiPass    string
+	apiVersion string
+}
+
+type apiConfigGetter interface {
+	GetAPIConfig() (*storageosAPIConfig, error)
+}
+
+type namespacedSecret struct {
+	namespace  string
+	secretName string
+	host       volume.VolumeHost
+}
+
+// GetAPIConfig checks the namespace for the api config.
+func (a *namespacedSecret) GetAPIConfig() (*storageosAPIConfig, error) {
+
+	c := &storageosAPIConfig{}
+
+	kubeClient := a.host.GetKubeClient()
+	if kubeClient == nil {
+		return nil, fmt.Errorf("Cannot get kube client")
+	}
+
+	keys, err := kubeClient.Core().Secrets(a.namespace).Get(a.secretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("Couldn't get secret %v/%v", a.namespace, a.secretName)
+	}
+	for name, data := range keys.Data {
+		if name == "apiAddress" {
+			c.apiAddr = string(data)
+		}
+		if name == "apiUsername" {
+			c.apiUser = string(data)
+		}
+		if name == "apiPassword" {
+			c.apiPass = string(data)
+		}
+		if name == "apiVersion" {
+			c.apiVersion = string(data)
+		}
+	}
+	return c, nil
+}
+
+func (s *storageosAPIConfig) SetAPIConfig(host volume.VolumeHost, namespace string) (string, error) {
+	kubeClient := host.GetKubeClient()
+	if kubeClient == nil {
+		return "", fmt.Errorf("Cannot get kube client")
+	}
+	secretName := "storageos-api-" + s.apiUser + "-secret"
+	secret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      secretName,
+		},
+		Data: map[string][]byte{
+			"apiAddress":  []byte(s.apiAddr),
+			"apiUsername": []byte(s.apiUser),
+			"apiPassword": []byte(s.apiPass),
+			"apiVersion":  []byte(s.apiVersion),
+		},
+		Type: "kubernetes.io/storageos",
+	}
+	_, err := kubeClient.Core().Secrets(namespace).Create(secret)
```

Comment:
this is unexpected... how is this used, and what component is expecting to have permission to create secrets in arbitrary namespaces?